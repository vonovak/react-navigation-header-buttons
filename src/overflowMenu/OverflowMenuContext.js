// @flow
import * as React from 'react';
import { Dimensions, Platform } from 'react-native';
import { Menu } from './vendor/Menu';
import { getSpaceAboveMenu } from './statusBarUtils';

export type ToggleMenuParam = ?{|
  elements: React.ChildrenArray<any>,
  x: number,
  y: number,
|};

export const OVERFLOW_TOP = 15;

const warn = () => {
  // the noop value will be replaced by OverflowMenuProvider rendered in app root
  console.warn(
    'It seems like you tried to open the overflow menu using the overflowMenuPressHandlerDropdownMenu' +
      ' - which is the default handler on android - but you forgot to wrap your root component in <OverflowMenuProvider />.' +
      'Please check the installation instructions in the react-navigation-header-buttons readme!'
  );
};
export const OverflowMenuContext = React.createContext<(ToggleMenuParam) => void>(warn);

type Props = {|
  children: React.Element<any>,
|};

export const OverflowMenuProvider = ({ children }: Props) => {
  const [visible, setVisible] = React.useState(false);
  const [position, setPosition] = React.useState({ x: Dimensions.get('window').width - 10, y: 40 });
  const [elements, setElements] = React.useState(null);

  const hideMenu = React.useCallback(() => {
    setVisible(false);
  }, []);

  const toggleMenu = React.useCallback((params: ToggleMenuParam) => {
    setVisible((prevVisible) => !prevVisible);
    setElements(params?.elements || []);
    if (params) {
      const { x, y } = params;
      const heightApprox = getSpaceAboveMenu();
      const extraDelta = Platform.select({
        android: heightApprox,
        default: OVERFLOW_TOP,
      });
      setPosition({ x, y: y + extraDelta });
    }
  }, []);

  return (
    <OverflowMenuContext.Provider value={toggleMenu}>
      {React.Children.only(children)}
      <Menu visible={visible} onDismiss={hideMenu} anchor={position}>
        {elements}
      </Menu>
    </OverflowMenuContext.Provider>
  );
};
