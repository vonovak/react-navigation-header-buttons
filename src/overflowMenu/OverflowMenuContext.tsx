import { Dimensions } from 'react-native';
import { getDefaultSpaceAboveMenu } from './statusBarUtils';
import { Menu } from './vendor/Menu';
import { useTheme } from '@react-navigation/native';
import {
  createContext,
  ReactElement,
  ReactNode,
  useCallback,
  useContext,
  useMemo,
  useState,
} from 'react';
import * as React from 'react';

import { ButtonsExtraMarginContext } from '../ButtonsWrapper';

export type ToggleMenuParam = {
  elements: React.ReactNode;
  x: number;
  y: number;
};

export const OVERFLOW_TOP = 15;

const warn = () => {
  // the noop value will be replaced by HeaderButtonsProvider rendered in app root
  console.warn(
    'It seems like you tried to open the overflow menu using the overflowMenuPressHandlerDropdownMenu' +
      ' - which is the default handler on android - but you forgot to wrap your root component in <HeaderButtonsProvider />.' +
      'Please check the installation instructions in the react-navigation-header-buttons readme :)'
  );
};
const OverflowMenuContext = createContext<{
  toggleMenu: (params?: ToggleMenuParam) => void;
}>({
  toggleMenu: warn,
});

export const useOverflowMenu = () => useContext(OverflowMenuContext);

type Props = {
  children: ReactElement;
  stackType: 'native' | 'js';
  spaceAboveMenu?: number;
};

export const HeaderButtonsProvider = ({
  children,
  spaceAboveMenu,
  stackType,
}: Props) => {
  const [menuState, setMenuState] = useState({
    visible: false,
    position: {
      x: Dimensions.get('window').width - 10,
      y: 40,
    },
    elements: null as ReactNode | null,
  });

  const {
    colors: { card },
  } = useTheme();

  const hideMenu = useCallback(() => {
    setMenuState((prevState) => ({
      ...prevState,
      visible: false,
    }));
  }, []);

  const toggleMenu = useCallback(
    (params?: ToggleMenuParam) => {
      const extraDelta = spaceAboveMenu ?? getDefaultSpaceAboveMenu();

      setMenuState((prevState) => {
        const position = params
          ? { x: params.x, y: params.y + extraDelta }
          : prevState.position;
        const elements = params ? params.elements : prevState.elements;

        return {
          ...prevState,
          position,
          elements,
          visible: !prevState.visible,
        };
      });
    },
    [spaceAboveMenu]
  );

  const { visible, position, elements } = menuState;
  const value = useMemo(() => ({ toggleMenu }), [toggleMenu]);
  const extraMarginValue =
    stackType === 'native' ? 'alreadyHandled' : 'toBeHandled';

  return (
    <OverflowMenuContext.Provider value={value}>
      <ButtonsExtraMarginContext.Provider value={extraMarginValue}>
        {React.Children.only(children)}
        <Menu
          visible={visible}
          onDismiss={hideMenu}
          anchor={position}
          contentStyle={{ backgroundColor: card }}
        >
          {elements}
        </Menu>
      </ButtonsExtraMarginContext.Provider>
    </OverflowMenuContext.Provider>
  );
};
