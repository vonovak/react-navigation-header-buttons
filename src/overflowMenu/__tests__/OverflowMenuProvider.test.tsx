import { HeaderButtonsProvider, useOverflowMenu } from '../OverflowMenuContext';
import { Text, View } from 'react-native';
import {
  render,
  fireEvent,
  waitForElementToBeRemoved,
} from '@testing-library/react-native';
import React from 'react';

describe('HeaderButtonsProvider renders', () => {
  it('only the child when menu is not shown', () => {
    const { toJSON } = render(
      <HeaderButtonsProvider stackType={'js'}>
        <Text>hello</Text>
      </HeaderButtonsProvider>
    );
    expect(toJSON()).toMatchInlineSnapshot(`
      <Text>
        hello
      </Text>
    `);
  });

  it('the child and the menu item, when it is displayed by calling presentMenu() that is provided by OverflowMenuContext', async () => {
    const menuItemLabel = 'overflow menu item';
    const showMenuLabel = 'show menu';
    const hideMenuLabel = 'hide menu';

    const ButtonThatShowsMenu = () => {
      const { presentMenu } = useOverflowMenu();
      return (
        <Text
          onPress={() => {
            presentMenu({
              elements: <Text key="1">{menuItemLabel}</Text>,
              x: 0,
              y: 0,
            });
          }}
        >
          {showMenuLabel}
        </Text>
      );
    };
    const ButtonThatHidesMenu = () => {
      const { closeMenu } = useOverflowMenu();
      return (
        <Text
          onPress={() => {
            closeMenu();
          }}
        >
          {hideMenuLabel}
        </Text>
      );
    };
    const { toJSON, getByText } = render(
      <HeaderButtonsProvider stackType={'js'}>
        <View>
          <ButtonThatShowsMenu />
          <ButtonThatHidesMenu />
        </View>
      </HeaderButtonsProvider>
    );

    const presentMenu = () => fireEvent.press(getByText(showMenuLabel));
    const hideMenu = () => fireEvent.press(getByText(hideMenuLabel));

    const beforeShown = toJSON();
    presentMenu();
    expect(toJSON()).toMatchSnapshot();
    getByText(menuItemLabel);
    hideMenu();

    const waitForMenuToHide = async () => {
      await waitForElementToBeRemoved(() => getByText(menuItemLabel), {
        timeout: 800,
      });
    };
    await waitForMenuToHide();

    const afterHidden = toJSON();
    expect(afterHidden).toStrictEqual(beforeShown);
  });
});
