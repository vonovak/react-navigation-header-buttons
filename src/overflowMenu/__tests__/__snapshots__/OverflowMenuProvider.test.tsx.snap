// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderButtonsProvider renders the child and the menu item, when it is displayed by calling presentMenu() that is provided by OverflowMenuContext 1`] = `
[
  <View>
    <Text
      onPress={[Function]}
    >
      show menu
    </Text>
    <Text
      onPress={[Function]}
    >
      hide menu
    </Text>
  </View>,
  <View
    collapsable={false}
    style={
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      accessibilityLabel="Close menu"
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
    <View
      accessibilityViewIsModal={true}
      collapsable={false}
      onAccessibilityEscape={[Function]}
      pointerEvents="box-none"
      style={
        [
          {
            "position": "absolute",
          },
          {
            "left": 8,
            "top": 8,
          },
          undefined,
        ]
      }
    >
      <View
        collapsable={false}
        style={
          {
            "transform": [
              {
                "translateX": -0,
              },
              {
                "translateY": -0,
              },
            ],
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "rgb(255, 255, 255)",
              "borderRadius": 5,
              "opacity": 0,
              "paddingVertical": 8,
              "shadowColor": "black",
              "shadowOffset": {
                "height": 7,
                "width": 0,
              },
              "shadowOpacity": 0.24,
              "shadowRadius": 8,
              "transform": [
                {
                  "scaleX": 0,
                },
                {
                  "scaleY": 0,
                },
              ],
            }
          }
        >
          <Text>
            overflow menu item
          </Text>
        </View>
      </View>
    </View>
  </View>,
]
`;
